arch/x86/core/intel64/locore.S

```diff
@@ -290,22 +307,66 @@ __resume:
 #define EXCEPT_CODE(nr) vector_ ## nr: pushq $nr; jmp except
 #define EXCEPT(nr)	vector_ ## nr: pushq $0; pushq $nr; jmp except
 
+/*
+ * When we arrive at 'except' from one of the EXCEPT(X) stubs,
+ * we're on the exception stack with irqs unlocked (or the trampoline stack
+ * with irqs locked if KPTI is enabled) and it contains:
+ *
+ *   SS
+ *   RSP
+ *   RFLAGS
+ *   CS
+ *   RIP
+ *   Error Code if pushed by CPU, else 0
+ *   Vector number <- RSP points here
+ *
+ */
+
 except: /*
 	 * finish struct NANO_ESF on stack. 'vector' .. 'ss' are
 	 * already there from hardware trap and EXCEPT_*() stub.
 	 */
+
+	pushq %r15
+
 #ifdef CONFIG_USERSPACE
-	/* Swap GS register values if we came in from user mode */
-	testb $0x3, 24(%rsp)
+	/* Swap GS register values and page tables if we came from user mode */
+	testb $0x3, 32(%rsp)
 	jz 1f
 	swapgs
+#ifdef CONFIG_X86_KPTI
+	/* Load kernel's page table */
+	movq $z_x86_kernel_ptables, %r15
+	movq %r15, %cr3
+#endif /* CONFIG_X86_KPTI */
 1:
 #ifdef CONFIG_X86_BOUNDS_CHECK_BYPASS_MITIGATION
 	/* swapgs variant of Spectre V1. Disable speculation past this point */
 	lfence
 #endif /* CONFIG_X86_BOUNDS_CHECK_BYPASS_MITIGATION */
+#ifdef CONFIG_X86_KPTI
+	/* Save old trampoline stack pointer in R15 */
+	movq %rsp, %r15
+
+	/* Switch to the exception stack */
+	movq %gs:__x86_tss64_t_ist7_OFFSET, %rsp
+
+	/* Transplant trampoline stack contents */
+	pushq 56(%r15)	/* SS */
+	pushq 48(%r15)	/* RSP */
+	pushq 40(%r15)	/* RFLAGS */
+	pushq 32(%r15)	/* CS */
+	pushq 24(%r15)	/* RIP */
+	pushq 16(%r15)	/* Error code */
+	pushq 8(%r15)	/* Vector */
+	pushq (%r15)	/* Stashed R15 */
```

Comment:
note: this overwrites any lingering traces of the thread's register contents, but not its execution context (vector through SS)